### PR TITLE
Refactor executors

### PR DESCRIPTION
**Pull request by rabernat**
### Overview

This is a major refactor of the internals of rechunker. I have changed the interface between the executors and the rechunking plan by _adding some new types_. The executors all now accept something called a `ParallelPipelines` object. The hierarchy of types looks like this

```
ParallelPipelines = Iterable[MultiStagePipeline]
MultiStagePipeline = Iterable[Stage]
class Stage(NamedTuple):
    func: Callable
    map_args: Optional[Iterable] = None
```

Stages contain a single function that is mapped across many inputs (e.g. a single copy operation). MultiStagePipelienes contain multiple Stages (e.g. copy source to intermediate, then intermediate to target). ParallelPipelines contain several MultiStagePipelines that can be executed in parallel.

The `rechunk` function now contains a line called `pipelines = specs_to_pipelines(copy_spec)` which translates a list of `CopySpec`s to a `ParallelPipelines`. This function contains all of the logic about how to execute a copy operation. All the executors needs to do is know how to execute a generic `ParallelPipelines`.

### Motivation

The abstractions we have created in rechunker, which allow many different distributed execution engines to be used for the same computation, are very useful and cool. The underlying motivation for this refactor was to make the executors more general, such that they can be used in other projects (e.g. Pangeo Forge). This is accomplished by decoupling the details of the rechunking operation as currently implemented from the `Executor` class.

### Pros

Beyond the motivation above, this approach has a couple of major benefits:
- Simplify Executor code
- Allow all Executors to accept Dask arrays 🎉  
- Allow read consolidation for Dask array sources (see e.g. #75)
- Enable future enhancements (e.g. add additional stages, like delete intermediate, consolidate metadata, etc.)

### Cons

But there are some downsides:
- Additional code complexity (more levels of abstraction)
- Possible performance impacts (have not done benchmarking comparison yet)
- Possible impacts on graph size / complexity (have not checked yet)

### Todo

- [x] Python Executor
- [x] Dask Exectuor
- [x] Prefect Exectuor
- [ ] Beam Exector
- [ ] PyWren Executor
- [ ] Additional tests?
- [ ] Documentation

At this point I would love to get a preliminary review from anyone who is interested.

### CODE REVIEW

**Comment by shoyer** [comment]
I like this general idea! My main concern is that this would preclude the option to avoid using intermediate arrays in favor of executor-native groupby operations, e.g., like what is sketched out in https://github.com/pangeo-data/rechunker/pull/36 for Beam. In principle, avoiding the intermediate copy could be up to twice as fast, if an executor like Beam or Spark manages to hold all the intermediate values in memory instead of dumping to disk.

**Comment by rabernat** [comment]
Good point Stephan. I think there is a simple resolution; we make the CopySpec -> Pipeline translation optional and allow executors to natively work on CopySpecs if they prefer. Stand by for an update to implement this.

**Comment by shoyer** [comment]
> Good point Stephan. I think there is a simple resolution; we make the CopySpec -> Pipeline translation optional and allow executors to natively work on CopySpecs if they prefer. Stand by for an update to implement this.

Sounds good to me!

File: rechunker/executors/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    return [spec_to_pipeline(spec) for spec in specs]
+
+
+# TODO: understand how to make these type annotations work with a mixin
+T = TypeVar("T")
+
+
+class CopySpecToPipelinesMixin:
+    # This signature doesn't work as a mixin because we don't know what type T is
+    def prepare_plan(self, specs: Iterable[CopySpec]) -> T:
```
Comment (rabernat) [inline rechunker/executors/pipeline.py:76]: @shoyer - I could use your advice on how to fix the type hints here. I am trying to use a mixin to transform `PipelineExecutor` to a `CopySpecExecutor`. The problem is that this mixin doesn't know what type `T`. This goes beyond my type hinting abilities...

Comment (rabernat) [reply rechunker/executors/pipeline.py:76]: Another ping on @shoyer for a little help with type hints.

**Comment by rabernat** [comment]
For context, the tests are currently failing due to the pre-commit mypy failure described above. That's the only blocker here.

File: rechunker/executors/pipeline.py (hunk 1 above)
Comment (shoyer) [reply rechunker/executors/pipeline.py:76]: I haven't tried to use type hints with mixins before. StackOverflow suggests a few approaches that may work: https://stackoverflow.com/questions/51930339/how-do-i-correctly-add-type-hints-to-mixin-classes

**Comment by rabernat** [comment]
I just put this through its paces on google cloud, and I'm satisfied it is working ok for real world use cases. I'm going to merge. It would be great if other users (maybe @rsignell-usgs?) could take this for a spin by installing rechunker from github master.

**Pull request merged**